File: 1402-CountSquareSubmatricesWithAllOnes/1402-CountSquareSubmatricesWithAllOnes.cpp
// Last updated: 7/24/2025, 8:27:22 AM
class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m+1,vector<int>(n+1,0));
        int res = 0;
        for(int i=1;i<=m;i++){
            for(int j=1;j<=n;j++){
                if(matrix[i-1][j-1]){
                    dp[i][j] = min(min(dp[i-1][j-1],dp[i-1][j]),dp[i][j-1]) + 1;
                    res += dp[i][j];
                }
            }
        }
        return res;
    }
};

File: 1402-CountSquareSubmatricesWithAllOnes/1402-CountSquareSubmatricesWithAllOnes.py
# Last updated: 7/24/2025, 8:27:27 AM
class Solution(object):
    def countSquares(self, matrix):
        m, n = len(matrix), len(matrix[0])
        dp = [[0] * (n+1) for _ in range(m+1)] # new int[m+1][n+1];
        ans = 0
        for i in range(m):
            for j in range(n):
                if matrix[i][j]:
                    dp[i+1][j+1] = min(dp[i][j+1], dp[i+1][j], dp[i][j]) + 1
                    ans += dp[i+1][j+1]
        return ans